### ISSUE TITLE
Better error handling and error messages

### ISSUE THREAD

**Issue by tjysdsg** [tjysdsg/tan#26]
- [x] Error handling shouldn't be done using C++ exceptions
- [x] Sometimes error messages are unclear, and the location of errors in source files is not intuitively formated

**Comment by tjysdsg** [comment]
Fix in 47c5c5c63d3976293bc1a700a52b52b50d32c2b7 and 082e1871c44c48276b416b3712149f38a8bb6197, but need further testing

**Comment by tjysdsg** [comment]
Many commits later, the error reporting have gotten better. BUT, I think it would be better if the position indicator can point out the start and the end of the expression that causes the error.

For example, the following error message only points out the start of the expression
```
fn hello(var name: str) : void {
              ^
```
while this gives the user a better understanding of what's wrong:
```
fn hello(var name: str) : void {
             ^^^^^^^^
```
